### ISSUE TITLE
Increment operator not recognised as invoking get accessor

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#43191]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/977932/increment-operator-not-recognised-as-invoking-get.html)._

---
On a property with a get and set accessor, I am being told that:

>Private property 'MovementController.Frame' can be converted to a method as its get accessor is never invoked.

However, if I remove the get accessor from the property, I get this error when I try to use the increment operator on the property:

>The property or indexer 'MovementController.Frame' cannot be used in this context because it lacks the get accessor

Is this a bug?

---
#### Visual Studio Feedback System on 4/7/2020, 04:19 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by zaytsev-victor** [comment]
```
// Compound assignment or increment whose value is being dropped (parent is an expression statement)
// is treated as a Write as the value was never actually 'read' in a way that is observable.
//
// Consider the following example:
//      class C
//      {
//          private int _f1 = 0, _f2 = 0;
//          public void M1() { _f1++; }
//          public int M2() { return _f2++; }
//      }
//
// Note that the increment operation '_f1++' is child of an expression statement, which drops the result of the increment.
// while the increment operation '_f2++' is child of a return statement, which uses the result of the increment.
// For the above test, '_f1' can be safely removed without affecting the semantics of the program, while '_f2' cannot be removed.
```

@sharwell should it be fixed?

**Comment by CyrusNajmabadi** [comment]
yes, this definitely should be fixed.

**Comment by slavanap** [comment]
I want to note, get accessor also not detected on usage of any operator on a property, like
`MyProp += 1;`

@jinujoseph why this issue is in a backlog? I have to mute it for all my projects with `.editorconfig` to have a clean build.

**Comment by CyrusNajmabadi** [comment]
> @jinujoseph why this issue is in a backlog?

This hasn't met our bar for being fixed in a specific release.  Note that this is marked help-wanted.  We would take a community PR to fix this if you wanted to provide it :)

**Comment by mavasani** [comment]
This is not a bug. The analysis is not ignoring the get accessor being used in the increment. It is doing further analysis that the result of get/set as part of the increment has no observable effect on the semantics of the code. As clearly explained in [this comment](https://github.com/dotnet/roslyn/issues/43191#issuecomment-619422918), especially the last statement of the comment, the property can indeed be removed.

**Comment by mavasani** [comment]
To further clarify, if ‘_f1’ was read anywhere else where it’s value was used, then it wouldn’t be a candidate for removal and IDE0052 would not be flagged.

**Comment by DoctorKrolic** [comment]
I would say, the message, that analizer produces, is really a bit confusing. The `get` accessor is used, however the value of the whole propery is not

**Comment by slavanap** [comment]
@mavasani set accessor may have a side effect that analyser disregards completely as shown in sample here inhttps://github.com/dotnet/roslyn/issues/53913

**Comment by mavasani** [comment]
> set accessor may have a side effect that analyser disregards completely as shown in sample here in https://github.com/dotnet/roslyn/issues/53913

Yes, this is known, but also a very weirdly written piece of code where property setter has such side effects but the better is never used. Almost 99% of cases where this fires will be code where the entire property can be safely removed. For the 1% case of setter with side effects, a source suppression with justification should be added.

**Comment by slavanap** [comment]
@mavasani please back up your numbers.

1. There's repository approach that usually imply side effects of storage the changes, as well as
2. properties in interfaces are pretty common, and their behavior is totally dependent on implementation. 
3. There's also WPF where you may want to shift object to specific offset and it would provide same warning.

In my opinion, for any property custom implementation that warning is misleading.

**Comment by CyrusNajmabadi** [comment]
> I would say, the message, that analizer produces, is really a bit confusing. The get accessor is used, however the value of the whole propery is not

Updating the message would be fine IMO.

**Comment by CyrusNajmabadi** [comment]
@slavanap this is why suppressions exist.  There are rarely hard rules that apply to everything.  If there were, we'd just make these compiler warnings/errors themselves.  The goal of analyzers is to flag the more-likely-than-not cases and let people adjust them.  If it turns out the code does depend on htat behavior, then disabling the analyzer or suppressing that location is the way to go.

**Comment by mavasani** [comment]
@slavanap I believe it should be the other way here as a bug is being raised to indicate false positive. Example cited in https://github.com/dotnet/roslyn/issues/53913 seems to be a made up piece of convoluted code. The role of a getter is to get the property value, and the setter is to set the new value. A setter going out of it's way to perform other operations and having side effects is badly written code - the correct way would be for moving all such operations to a separate method.

I agree with @DoctorKrolic and @CyrusNajmabadi here - we should update the message to be clear that there is indeed a code quality issue here, it may mean you can delete the entire property or refactor the code so that the side effects are moved outside of the property setter for the code to be more explicit and clear. If you still prefer that you have setters with unintuitive side effects, you should add a source suppression on the setter with a comment that clarifies that you indeed want to keep this non-recommended code pattern.

**Comment by slavanap** [comment]
@CyrusNajmabadi same question to you, if property may have a custom (not generated implementation), please back up your statement that such warning is "the more-likely-than-not case". From my experience, it's otherwise.

**Comment by mavasani** [comment]
Actually, looking at the diagnostic message in this issue description:

> Private property 'MovementController.Frame' can be converted to a method as its get accessor is never invoked.

I believe the message is not telling you that the property should be removed. It is indeed telling you there is something fishy about this property as it likely has side effects, and should be refactored to have the code with side effects moved to a method instead. If you disagree with this recommendation for specific instances of warnings, please add a source suppression at each site. If you disagree with this recommendation for all cases, feel free to disable the rule via editorconfig.

**Comment by slavanap** [comment]
@mavasani it's not constructive to make statements without samples or proofs or links to independent sources, especially when asked for. Meanwhile I leave sample below to think of

```
var entity = dbContext.Entities.AsQueryable().First();
entity.Counter++;
dbContext.Update(entity);
dbContext.SaveChanges();
```

**Comment by DoctorKrolic** [comment]
@slavanap This analizer is about private members, but you give an example, where you have access to `Counter` property outside your `Entity` class declaration

**Comment by mavasani** [comment]
> > Almost 99% of cases where this fires will be code where the entire property can be safely removed.
> 
> > @mavasani please back up your numbers.
> 
> > A setter going out of it's way to perform other operations and having side effects is badly written code
> 
> @mavasani it's not constructive to make statements without samples or proofs or links to independent sources, especially when asked for. Meanwhile I leave sample below to think of
> 
> ```
> var entity = dbContext.Entities.AsQueryable().First();
> entity.Counter++;
> dbContext.Update(entity);
> dbContext.SaveChanges();
> ```

@slavanap Your claim is that the analyzer warning is a false positive. For all the examples you have provided, the analyzer is recommending you to convert the property into a method to make it’s side effects causing setter behaviour be more explicit. We believe this is indeed a valid code quality recommendation for properties written in such way - we don’t feel developers should be writing setters with unintuitive side effects, it is better refactored to have such code be moved into a separate method. There is no false positive here. If you disagree with the suggestion, kindly add a source suppression or disable the rule for your repo.

**Comment by slavanap** [comment]
@mavasani 
I recall where I've faced it -- when I tried to mirror a property and/or have alternative private name.
My full sample - https://gist.github.com/slavanap/779dd0228025ba197937fd3fd4ea3b68

Simplified sample -- desire to have 2 property names to point to the same memory/storage, i.e.
```
class C {
  int A { get; set; }
  int B { get => A; set { A = value; } }
}
```
Having in mind that C# supports Unicode property names its a helpful approach to indicate parts of logic in different human languages.

I agree that such analyser warning is helpful for generated property getters and setters, but with customized setter, the setter is actually using the getter result, which the analyser marks as never being invoked. If I could I would disable the warning for properties with customized setters, but there's no such option is seems.

**Comment by CyrusNajmabadi** [comment]
@slavanap Yup, i would highly not recommend that.  That's extremely unintuitive and not at all what i would expect from properties.
